Block Support: Update border support to allow non-pixel units

aaronrobertshaw opened the pull request:
Part of: https://github.com/WordPress/gutenberg/issues/31337
Related: https://github.com/WordPress/gutenberg/pull/31641, https://github.com/WordPress/gutenberg/pull/31585

## Description
- Updates border block support feature to allow custom CSS units for width and radius
- Replaces RangeControl use with UnitControl to support custom units in prep for UI updates in https://github.com/WordPress/gutenberg/issues/31337
- Tweaks the step value depending on current field's CSS unit type
- Adds custom CSS for border block support hook controls (will be extended for UI updates)

Given border block support is still `__experimental` no deprecation has been added for the moment.

## How has this been tested?
Manually.

1. Checkout this PR and enable border block support via your `experimental-theme.json`. To save time enabling it for multiple block contexts, just update the `settings.defaults.border` flags.
2. Create a new post, add a group block and select it.
3. Expand the border controls in the sidebar. The border width and radius should now show a `UnitControl`.
4. Adjust the border values, save and confirm the block updates as expected both in the editor and frontend.
5. Opt into border support for the post tags block via its `block.json` file.
6. Reload the editor and add a post tags block along with some tags to fill it.
7. Select the post tags block, adjust border values, then save.
8. Refresh the post on the frontend and confirm this dynamic block renders the border correctly.

## Screenshots <!-- if applicable -->
<img src="https://user-images.githubusercontent.com/60436221/117091250-05e0f900-ad9e-11eb-97ad-bb0805e19821.gif" width="400" />

## Types of changes
New feature. 
Breaking Change (Border block support is still experimental however)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Update border support to allow non-pixel units

aaronrobertshaw commented:
@nosolosw Given the border block support feature is still experimental, can you confirm this change is ok to make without still handling previous unitless values? Or am I missing something?

stokesman commented:
I tried this and found the Button block doubles up on units.
![image](https://user-images.githubusercontent.com/9000376/117169544-079ac480-ad7e-11eb-9b16-2fa0b0bb3073.png)

Simple changes in the Buttons block can have it working. I'm not sure that there'd be no migration concern though. 

Otherwise, this worked well for me though I could only test it with the group block. For some reason, opting-in to support on the Post Tags block didn't work on my end.

Incidentally, it looks like this PR would resolve #29210.

aaronrobertshaw commented:
Thanks for testing @stokesman 

> I tried this and found the Button block doubles up on units.

The button block does skip the usual serialization of the block support styles for border radius. When it applies the radius manually it appended the `px` unit rather than rely on that to happen automatically. 

> Simple changes in the Buttons block can have it working. I'm not sure that there'd be no migration concern though.

You are correct it is just a simple change to that block. I'll update that as part of this PR.

Regarding the migration concerns, I believe the issue might only be for dynamic blocks. 

In static blocks, when the block support style was applied, it was just set to the number value and the px unit was applied automatically, so was in the saved content. Static blocks that had a border radius value without a specified unit would continue to function as they were. If the radius was adjusted the attribute value would then have a unit, which would be then saved into the content as expected. I'll test and confirm all of this.

Dynamic blocks had the `px` unit set when the style was generated in `border.php`. I could update the PHP side of this to check if the value contains a unit. If it doesn't, append the pixel unit as it did originally. 

Also, looking at other blocks that have skipped serialization of the block support style for border radius, the search block might be a concern. Depending on its layout, button inside vs outside, it will apply the border radius differently. It relied on the fact the value was only in pixels to calculate the adjustment. It might be better to address that via a separate PR once non-pixel values are allowed for the border radius.

aaronrobertshaw committed:
Fix button block application of border radius

Button block previously appended the `px` unit when manually applying the border radius style. This change removes that as the unit will be included in the block support attribute for border radius.

aaronrobertshaw committed:
Prevent empty string being saved to border attributes

aaronrobertshaw committed:
Handle border backwards compatibility for dynamic blocks

aaronrobertshaw committed:
Fix search block to work with non-pixel border radius units

aaronrobertshaw commented:
The fixes for the button and search blocks were relatively straightforward so I've ended up including them in this PR so there is no gap between this landing and those blocks working with non-pixel units.

The search block solution required inlining a style using `calc()`. If anyone sees any issues with this approach I can split that into its own PR.

I've searched for other blocks opting into border support and of those they all work for me without validation errors when loading a deprecated version in the editor or frontend. These blocks include; button, search, table, image & group.

glendaviesnz commented:
I tested this and from a functional prespective everything worked as advertised:
![borders](https://user-images.githubusercontent.com/3629020/117400045-f4425180-af55-11eb-8641-08159ac3460d.gif)

I will leave it to someone with more experience in the block supports to give it s full code review signoff.

aaronrobertshaw committed:
Update border support style tests for custom units

cgaits commented:
I would like to comment on this topic once more and express that, in the border drop-down menu there should be a NONE above solid.

aaronrobertshaw commented:
Thanks for the feedback @cgaits 👍 

This PR is only concerned with being able to allow custom CSS units for the border width and border radius block support.

There has been some discussion regarding a "none" option for the border style support on the issues relating to refining the border block support UI. https://github.com/WordPress/gutenberg/issues/31337 & https://github.com/WordPress/gutenberg/issues/31333.

The most recent proposal is that the border width be responsible for disabling the border i.e. a border width of `0` sets `border-style: none`. 

This will be addressed as part of https://github.com/WordPress/gutenberg/pull/31585

aaronrobertshaw commented:
Also, in case it interests @cgaits, the original UI (which is the version in this PR) does actually have a "none" option in the border style dropdown above "solid".

<img width="276" alt="Screen Shot 2021-05-21 at 12 04 33 pm" src="https://user-images.githubusercontent.com/60436221/119071401-14c9db80-ba2d-11eb-87d8-a306e596449f.png">

stokesman commented:
I linked #29210 since this PR happens to resolve it and want to suggest that the behavior (defaulting to empty radius value) is correct and should be maintained here.

ramonjd commented on packages/block-editor/src/hooks/border-width.js:
```diff
@@ -1,16 +1,17 @@
 import { cleanEmptyObject } from './utils';
 
 const MIN_BORDER_WIDTH = 0;
-const MAX_BORDER_WIDTH = 50;
```
Are we getting rid of max for any reason or because it doesn't play well with varying units?

aaronrobertshaw replied:
There were two main reasons behind me making that change. First, was the max was required to utilise the slider control in the original border radius support UI. Second, a max radius of `50px` wasn't very flexible at all and at worst simply didn't make sense in some contexts.

Given a max value wasn't required for the `UnitControl` and that the max would need to change based on context e.g. type of block, dimensions of the block etc, I saw only minor downside in removing the max. The rounded block style for the image block for example uses `border-radius: 9999px`.

I'm happy to iterate on this in future if we do want to constrain values more. We'd need some means in the block support mechanism to be able to provide the correct max for the context. It could be a flag within the block.json support configuration for example.

ramonjd commented on packages/block-editor/src/hooks/border.js:
```diff
@@ -16,7 +16,41 @@ import { BorderRadiusEdit } from './border-radius';
+		value: 'rem',
+		label: isWeb ? 'rem' : __( 'Relative to root font size (rem)' ),
+		default: '',
+	},
```
If there's a desire to reintroduce max values, maybe here's a decent spot for them?

aaronrobertshaw replied:
If we could get away with static max values, that would be a good place for them. 

I'm not sure how much sense it makes in the bigger picture. You can get some weird/creative shapes with complete freedom over the border-radius values. That said we generally try to encourage good behaviour by default so I'm by no means against a future iteration to add max values if we have a flexible strategy for them.

ramonjd replied:
All good, was asking out of curiosity. Your reasoning make sense to me. 👍

aaronrobertshaw replied:
Thanks, appreciate the sanity check 👍

ramonjd commented:
Tested according to instructions (on web only) in FF, Chrome and Safari. Group block border controls appear as expected in both editor and fronted for all units, and in the absence of a value.

Similarly,  the search block border radius functions as expected.

> Replaces RangeControl use with UnitControl to support custom units in prep for UI updates in #31337

Not a requirement for this PR, but just jotting it down for reference: now that the range control is absent, and there's no UI to indicate how one is able to adjust controls without the keyboard, would a cheap way be to add a cursor? E.g., using `cursor: ns-resize` on `ValueInput input`

aaronrobertshaw commented:
Thanks for testing @ramonjd !

> Not a requirement for this PR, but just jotting it down for reference: now that the range control is absent, and there's no UI to indicate how one is able to adjust controls without the keyboard, would a cheap way be to add a cursor? E.g., using cursor: ns-resize on ValueInput input

I do see a cursor but only after I commence dragging up/down on the control. I'm not sure how best we can improve that UX. Changing the default cursor would obscure that the field is keyboard editable. 

Either way, this is specific to the `UnitControl` component. An accessibility review could be done on it separately and that control iterated on based on its findings.

oandregal commented on packages/block-editor/src/hooks/border.js:
```diff
@@ -16,7 +16,41 @@ import { BorderRadiusEdit } from './border-radius';
+const isWeb = Platform.OS === 'web';
+
 export const BORDER_SUPPORT_KEY = '__experimentalBorder';
+export const CSS_UNITS = [
```
Not for this PR, but wanted to share some thoughts about this, in case someone here has bandwidth to act on them. I find asking the consumers of the `UnitControl` to pass a list of unit definitions creates a couple of issues:

- strings to translate: asking consumers of the `UnitControl` to redefine units (they're already defined here) creates the same multiple entries for folk to translate, which are the same.
- maintanibility: all consumers need to be changed when we want to change a label, for example.

It'd be better if the component only asked consumers to pass the allowed units and its defaults.

---

For this PR: I see that the units defined [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/unit-control/utils.js#L17-L17) also have a `a11ylabel`, which this list doesn't have. Do we need it?

aaronrobertshaw replied:
I've added the `a11yLabel` properties to the `CSS_UNITS` config used by border support. 

It appears they are only used by the native UnitControl at the moment and they get retrieved from the UnitControl's `ALL_CSS_UNITS` config in `utils.js`. I still added them here for consistency and to withstand any changes that might happen to the UnitControl before your suggestion above can be implemented. I figure it being consistent will make that a smoother process as well.

On the topic of updating the UnitControl to only ask consumers to pass allowed units and defaults, that sounds good to me. I'll make a note of this in my backlog and try and free up some bandwidth to address it.

aaronrobertshaw replied:
I've created an issue for the proposed UnitControl changes. https://github.com/WordPress/gutenberg/issues/32478

aaronrobertshaw replied:
Turns out this has already been taken care of https://github.com/WordPress/gutenberg/issues/32478#issuecomment-855705065.

There's a PR up to update the use of `CSS_UNITS` here and in the letter-spacing hook (https://github.com/WordPress/gutenberg/pull/32482)

oandregal commented on packages/block-editor/src/hooks/border-width.js:
```diff
@@ -24,27 +25,40 @@ export const BorderWidthEdit = ( props ) => {
 			},
 		};
 
+		if ( newWidth === undefined || newWidth === '' ) {
```
As per a few lines below, we already clean the object in all cases, why do we need this?

oandregal replied:
Or rather: why do we clean the object? Is it to remove the width when it has a falsy value? If so, we don't need to do it always.

aaronrobertshaw replied:
Thanks for catching that one. An oversight while resolving a conflict during a rebase.

The intention is to allow the `0` border width value. I've removed the erroneous call to `cleanEmptyObject`.

oandregal reviewed (approved):
I've left a couple of comments that I'd like us to adddress. They're minor and the feature works as expected, so I'm pre-approving. Thanks!

aaronrobertshaw committed:
Remove extraneous cleanEmptyObject call

aaronrobertshaw committed:
Add a11yLabels for border CSS Units

aaronrobertshaw commented:
I've made the requested changes, rebased and re-tested. Once the e2e tests pass, this should be right to merge, then https://github.com/WordPress/gutenberg/pull/31641 and https://github.com/WordPress/gutenberg/pull/31585 can be rebased as well.
